java/org/apache/catalina/util/TimeBucketCounter.java

```diff
@@ -0,0 +1,217 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.catalina.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * this class maintains a thread safe hash map that has timestamp-based buckets
+ * followed by a string for a key, and a counter for a value. each time the
+ * increment() method is called it adds the key if it does not exist, increments
+ * its value and returns it.
+ *
+ * a maintenance thread cleans up keys that are prefixed by previous timestamp
+ * buckets.
+ */
+public class TimeBucketCounter {
+
+    /**
+     * Map to hold the buckets
+     */
+    private final ConcurrentHashMap<String, AtomicInteger> map = new ConcurrentHashMap<>();
+
+    /**
+     * Milliseconds bucket size as a Power of 2 for bit shift math, e.g.
+     * 16 for 65_536ms which is about 1:05 minute
+     */
+    private final int numBits;
+
+    /**
+     * ratio of actual duration to config duration
+     */
+    private final double ratio;
+
+    /**
+     * flag for the maintenance thread
+     */
+    volatile boolean isRunning = false;
+
+    /**
+     *
+     * @param bucketDuration duration in seconds, e.g. for 1 minute pass 60
+     */
+    public TimeBucketCounter(int bucketDuration) {
+
+        int durationMillis = bucketDuration * 1000;
+
+        int bits = 0;
+        int pof2 = nextPowerOf2(durationMillis);
+        int bitCheck = pof2;
+        while (bitCheck > 1) {
+            bitCheck = pof2 >> ++bits;
+        }
+
+        this.numBits = bits;
+
+        this.ratio = ratioToPowerOf2(durationMillis);
+
+        int cleanupsPerBucketDuration = (durationMillis >= 60_000) ? 6 : 3;
+        Thread mt = new MaintenanceThread(durationMillis / cleanupsPerBucketDuration);
+        mt.start();
+    }
+
+    /**
+     * increments the counter for the passed identifier in the current time
+     * bucket and returns the new value
+     *
+     * @param identifier an identifier for which we want to maintain count, e.g. IP Address
+     * @return the count within the current time bucket
+     */
+    public final int increment(String identifier) {
+        String key = getCurrentBucketPrefix() + "-" + identifier;
+        AtomicInteger ai = map.computeIfAbsent(key, v -> new AtomicInteger());
+        return ai.incrementAndGet();
```

Comment:
is it worth using `updateAndGet` or accumulate flavors? here you can end up enabling requests if you do a full rotation having enough requests in the configured window

Reply:
One idea that I had was to allow to configure certain paths with a different "weight", rather than configure a different Filter for different path, e.g. while requests to the homepage are counted as 1, a request for a login page might count as 10.

That way, you can configure the Filter for 300 requests per minute, for example, which will allow for 300 requests to most of the site, but only 30 requests to the login page, mitigating brute force attacks that try to guess a password.

If we do that then `updateAndGet()` might be more applicable, but again, I want to introduce this filter and get some feedback before adding more features to it.

Also, we have another thread on the PR about back porting this feature, and if we decide to go all the way back to Tomcat 8.5 then this would be another one of the things that needs to be rewritten to run on Java 7.

Reply:
this can easily be mimic-ed for java 7 but point is right now you can disable the rate limit if you get enough requests. If you take the extreme case for a rate limiter - DDoS - then you can just not reach this goal so think it is worth ensuring you don't go > max+1.

Reply:
@rmannibucau Perhaps I'm missing something.  Can you please explain how `updateAndGet()` is better than `incrementAndGet()`?  

Looking at the implementation of both they look very similar other than the fact that `updateAndGet()` would have to call `updateFunction.applyAsInt()` which would add unnecessary overhead.  Both implementations call `weakCompareAndSetInt()`.

Reply:
overhead will be 0 since it will be inlined at some point but the function enables to set the max instead of incrementing and getting the issue i mention. Alternatively you can workaround it by doing `if (incrAndGet > max) v.set(max+1)`, it is no more atomic but likely saner than nothing.

Reply:
@rmannibucau But you still didn't explain how that would be better than the current code and didn't post any references.  We should always push better code when possible, but can you explain what problem that change would solve and how?  Thank you.

Reply:
@isapir well I tried to explain in your case if you have a long window (for business reason, it is a common case) you can overflow "int.max" and start allowing requests which should be forbidden whereas using an atomic "update" instead of blind increment logic enables to not increment when you already are higher than the max (therefore my proposal to reset the counter to max+1 when increment > max+1). This is easy to get in an unit test even if "slow".

Reply:
To exceed Integer.MAX_VALUE (2,147,483,647) in a day, for example, one would need to hit the server at the rate of almost 25,000 Requests per Second.  Handling cases like that are not a use case for this filter.

Reply:
Ok-ish but so should maybe be written as a limitation in the filter class javadoc cause rate limiting is way more generic than current impl (which is ok but reading the doc it is not obvious IMHO).